### DISCUSSION TITLE
INA3221 / INA226 as Power Monitor?

### DISCUSSION THREAD

**Discussion by Battlecake91** [arendst/Tasmota#19248] [Q&A]
Hello everybody,

I try to measure as many consumers as possible to get a detailed overview.

Now I have a INA3221 with a ESP32-C3 and it works fine and shows me the actual values like power. 

But is it possible to Monitor the Power like other ICs (HW8032 for example) to see the today/yesterday/total values?
The INA3221 measures my router, server and switch - I guess the dynamic of the current is too high to read it over MQTT and integrate it.

**Comment by sfromis** [comment]
No, Tasmota only supports those chips to check the current power values, not for cumulative energy monitoring.

If you want to integrate values, you could create some Berry code to read values frequently, and do a calculation, but that would still not be exact for fast fluctuations or short spikes.

**Comment by barbudor** [comment]
Currently drivers for those are of SENSOR type
It is possible to create alternative version of the driver of type ENERGY
That would be a, somehow, different driver that would have to be enabled through a different #define
Most low level code would be the same but still due to the different driver class, it would need to be in the energy folder

Actually it may be possible to get only a cover energy driver which gets values from the sensor driver and push to Tasmota the energy values. 

It needs someone to get interrest in the topic to make such driver

**Reply by barbudor** [reply]
Berry does give access to energy driver features, so it should be possible to create such driver in berry

**Reply by pkkrusty** [reply]
Using InfluxDB to gather the instantaneous data and Integrating it in Grafana should get you the totals you need. A few more steps, but is possible. Can get to 10 second resolution with current influx driver.

**Reply by sfromis** [reply]
I'd expect the OP to be right about samples at 10 second intervals to be way too slow to deal with dynamic fluctuations in energy draw, except as an assumption of things averaging out with individual samples not representative of the interval. In principle, you should have sampling no slower than half the width of the spikes.

**Reply by Battlecake91** [reply]
thanks a lot barbudor for this information, I will playing arround with this hint. 

And sfromis is right - 10 seconds per sample is too high, the important measurement goes to a pico-psu in my server, I guess there are short load jumps for a few milliseconds. With a bigger capacitor on the input from the INA3221 it should be possible to integrate it a bit, but a time constant of about 10 seconds isn't possible at all.
I think 10 samples per second is the minimum amount to gather enough accuracy.

**Comment by rin67630** [comment]
> Hello everybody,
> 
> I try to measure as many consumers as possible to get a detailed overview.
> 
> Now I have a INA3221 with a ESP32-C3 and it works fine and shows me the actual values like power.
> 
> But is it possible to Monitor the Power like other ICs (HW8032 for example) to see the today/yesterday/total values? The INA3221 measures my router, server and switch - I guess the dynamic of the current is too high to read it over MQTT and integrate it.

As others mentioned, AC power monitoring do the integration by hardware. If you want to process it in software with an acceptable precision, you need a very fast processing, ways faster than Tasmota provides.
You may want to have a look an my -non-Tasmota development https://github.com/rin67630/Victron_VE_on_Steroids in the variant "Shunt logger. The signal is processed at 125mS and the ESP device makes the long-term integration.
I am currently working at an integration to provide Tasmota-compatible MQTT messages.

Alternatively you may use PZEM17 devices that integrate DC values and have a Tasmota setting.
